### PR TITLE
[SPARK-26560][SQL]:Repeat select on HiveUDF fails

### PR DESCRIPTION
**Pull request by nivo091**
## What changes were proposed in this pull request?

The classloader of spark-shell is IMainsTranslatingClassLoader which is loaded from IMain (scala.tools.nsc.interpreter). But on the first select, it registers the function and loads the hiveUDF jar to sparkContext classloader which is NonClosableMutuableClassLoader. While selecting the function on the second time, function is already registered so it tries to fetch from IMainTranslatingClassLoader  which is giving analysis exception.

Changing of the classLoader of currentThread to NonClosableMutuableclassLoader will solve this issue.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Can you fix the PR title? It sounds like we have a problem in `SELECT` of SparkSQL. The problem is about Hive UDF in a specific case.

**Comment by HyukjinKwon** [comment]
Also, please describe how the current PR fixes the issue in the PR description.

**Comment by nivo091** [comment]
I have updated the PR description and title, please review. @HyukjinKwon

**Comment by nivo091** [comment]
@HyukjinKwon This issue happens only in Spark-shell, that's why I added in the title. Is that not required ?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 1]
```diff
@@ -66,51 +66,58 @@ private[sql] class HiveSessionCatalog(
...
-          val analysisException = new AnalysisException(errorMsg)
-          analysisException.setStackTrace(e.getStackTrace)
-          throw analysisException
-      }
-      udfExpr.getOrElse {
-        throw new AnalysisException(s"No handler for UDF/UDAF/UDTF '${clazz.getCanonicalName}'")
       }
     }
+    finally
+    {
```
Comment (sujith71955) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: pull up the flower bracket inline with finally

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 2]
```diff
@@ -66,51 +66,58 @@ private[sql] class HiveSessionCatalog(
...
-            new HiveFunctionWrapper(clazz.getName),
-            input,
-            isUDAFBridgeRequired = true))
-          udfExpr.get.dataType // Force it to check input data types.
-        } else if (classOf[GenericUDTF].isAssignableFrom(clazz)) {
-          udfExpr = Some(HiveGenericUDTF(name, new HiveFunctionWrapper(clazz.getName), input))
-          udfExpr.get.asInstanceOf[HiveGenericUDTF].elementSchema // Force it to check data types.
+    val originalClassLoader = Thread.currentThread().getContextClassLoader()
+    Thread.currentThread().setContextClassLoader(clazz.getClassLoader)
+    try {
```
Comment (sujith71955) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: is this try block is required? please check once again

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 3]
```diff
@@ -66,51 +66,58 @@ private[sql] class HiveSessionCatalog(
...
+      Try(super.makeFunctionExpression(name, clazz, input)).getOrElse {
+        var udfExpr: Option[Expression] = None
+        try {
```
Comment (sujith71955) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: is this try block is required now? please check once again

Comment (nivo091) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: we can remove this inside try and put the catch block to the outer try.

**Comment by sujith71955** [comment]
@nivo091  Please correct the PR title format  seems to be not in same as standard mentioned by the community.

**Comment by nivo091** [comment]
@sujith71955 : Handled, thanks

**Comment by sujith71955** [comment]
cc @srowen @HyukjinKwon

**Comment by nivo091** [comment]
cc @Xiao Li

**Comment by HyukjinKwon** [comment]
Can you make the PR description concise? It's difficult to read and follow. Why does the classloader change in the second time? Can you also point out the related codes?

**Comment by nivo091** [comment]
Spark-shell have the IMainTranslatingClassLoader which is loaded from IMain (scala.tools.nsc.interpreter).
For the first select of function, it loads the hiveUDF jar in the below line. 
https://github.com/apache/spark/blob/e402de5fd030cdc4150fda0755c7c636cad9619e/sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala#L169

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 4]
```diff
@@ -66,10 +66,11 @@ private[sql] class HiveSessionCatalog(
       name: String,
       clazz: Class[_],
       input: Seq[Expression]): Expression = {
-
-    Try(super.makeFunctionExpression(name, clazz, input)).getOrElse {
-      var udfExpr: Option[Expression] = None
-      try {
+    val originalClassLoader = Thread.currentThread().getContextClassLoader()
```
Comment (HeartSaVioR) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: It doesn't need to modify original code but just switch context classloader. Then, use `Utils.withContextClassLoader` and just pass original code into fn.

**Comment by HeartSaVioR** [comment]
Btw, please follow the template of PR description. And we describe PR title for what the patch fixes instead of what was the bug.

**Comment by HeartSaVioR** [comment]
I'm taking this over, as it's a bit old and the test should be added as well. Please refer #27025.
I've retained the commit to give proper credit. Thanks!

**Pull request closed without merging**